sw/simulator/mesonh/mesonh.py

```diff
@@ -0,0 +1,203 @@
+sys.path.append(PPRZ_SRC + "/sw/ext/pprzlink/lib/v1.0/python")
+
+from pprzlink.ivy import IvyMessagesInterface
+from pprzlink.message import PprzMessage
+
+M_IN_KM = 1000.
+
+atm = None
+origin = np.array([0, 0, 0, 0])
+scale = np.array([1., 1/M_IN_KM, 1/M_IN_KM, 1/M_IN_KM])
+
+#ivy_req = None
+#ivy_mes = None
+
+
+#nps_udp_socket = None
+#nps_udp_dest = None
+
+start_time = time.time()
+
+
+def get_wind(east, north, up):
+	t = time.time() - start_time
+	print("east :",east)
+	print("north :",north)
+	print("up :",up)
+	loc = np.array([t, up, east, north])
+	loc = loc*scale + origin
+	print("loc:",loc)
+	weast, wnorth, wup = atm.get_wind(loc)
+	return weast, wnorth, wup
+
+
+def ivy_request_callback(sender, msg, resp, *args, **kwargs):
+    """
+        Ivy Callback for Paparazzi Requests
+    """
+
+    if msg.msg_class == "ground" and msg.name == "WORLD_ENV_REQ":
+        return worldenv_cb(msg, resp)
+    else:
+        return None
+
+
+#def worldenv_cb(m, r):
+def worldenv_cb(ac_id, msg):
+    """
+        Callback for paparazzi WORLD_ENV requests
+    """
+    # request location (in meters)
+    east, north, up = float(msg.get_field(3)),\
+        float(msg.get_field(4)),\
+        float(msg.get_field(5))
+    up *= -1
+    # convert in km + translation with mesoNH origin
+    weast, wnorth, wup = get_wind(east, north, up)
+    print("wind_est:")
+    print(weast)
+    print(wnorth)
+    print(wup)
+    msg_back=PprzMessage("ground", "WORLD_ENV")
+    msg_back.set_value_by_name("wind_east",weast)
+    msg_back.set_value_by_name("wind_north",wnorth)
+    msg_back.set_value_by_name("wind_up",wup)
+    msg_back.set_value_by_name("ir_contrast",400)
+    msg_back.set_value_by_name("time_scale",1)
+    msg_back.set_value_by_name("gps_availability",1)
+    ivy.send(msg_back,None)
+
+
+def ivy_callback(acid, msg):
+    """
+    Ivy callback for non-request paparazzi messages.
+    
+    NPS_SPEED_POS message is used to now ac position and send wind
+    data to NPS
+    """
+    if msg.msg_class == "telemetry" and msg.name == "NPS_SPEED_POS":
+        north, east, down = float(msg.ltpp_x),\
+            float(msg.ltpp_y),\
+            float(msg.ltpp_z)
+        up = -down
+
+        weast, wnorth, wup = get_wind(east, north, up)
+
+        wup *= -1
+
+        send_wind_nps_udp(acid, weast, wnorth, wup)
+
+        # some debug print
+        ivy_callback.acid_print[acid] = "[{}] pos {:.2f} {:.2f} {:.2f} wind {:.2f} {:.2f} {:.2f}"\
+            .format(acid, east, north, up, weast, wnorth, wup)
+        if ivy_callback.last_print + 1 <= time.time():
+            for s in ivy_callback.acid_print.viewvalues():
+                print(s)
+            ivy_callback.last_print = time.time()
+            ivy_callback.acid_print.clear()
+
+
+ivy_callback.last_print = time.time()
+ivy_callback.acid_print = {}
+
+
+#def send_wind_nps_udp(acid, wnorth, weast, wup):
```

Comment:
Not sure anymore, this is a student work from last year. I thought this was the working version, but maybe not. Should I remove it or keep it as a reference until I fix it or find a better version (some colleagues from LAAS are also working on this) ?